google_guest_agent/hostname/hostname.go

```diff
@@ -0,0 +1,241 @@
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package hostname reconfigures the guest hostname (linux only) and fqdn as necessary. It will do so on a detected change to the metadata hostname, when a new interface is configured, or a new ip address is acquired. It does so by triggering the HostnameReconfigure event, which is also triggerable outside the guest agent through named pipes on windows and unix sockets on linux. All of this behavior is configurable in the guest agent configuration.
+package hostname
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/cfg"
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/command"
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/events"
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/events/metadata"
+	mds "github.com/GoogleCloudPlatform/guest-agent/metadata"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+var (
+	// ReconfigureHostnameCommand is the command id registered for hostname configuration
+	ReconfigureHostnameCommand = "agent.hostname.reconfigurehostname"
+	disallowedConfigurations   = map[string]bool{"": true, "metadata.google.internal": true}
+	hostnameFqdnMu             = new(sync.Mutex)
+	lastHostname               string //As retrieved from MDS
+	lastFqdn                   string //As retrieved from MDS
+)
+
+// ReconfigureHostnameRequest is the structure of requests to the ReconfigureHostnameCommand
+// Regardless of request options, guest agent will only set the hostname or fqdn if enabled in the MDS.
+type ReconfigureHostnameRequest struct {
+	command.Request
+}
+
+// ReconfigureHostnameResponse is the structure of responses from the ReconfigureHostnameCommand
+// Status code meanings:
+// 0: everything ok
+// 1: error setting hostname
+// 2: error setting fqdn
+// 3: error setting hostname and fqdn
+type ReconfigureHostnameResponse struct {
+	command.Response
+	// Hostname is the hostname which was set. Empty if unset, either due to configuration or error.
+	Hostname string
+	// Fqdn is the hostname which was set. Empty if unset, either due to configuration or error.
+	Fqdn string
+}
+
+// Init registers a hostname command handler and subscribes to the metadata longpoll event if the user has enabled network interface setup and either hostname or fqdn management.
+func Init(ctx context.Context, mdsclient mds.MDSClientInterface) {
+	if cfg.Get().Unstable.SetFqdn || cfg.Get().Unstable.SetHostname {
```

Comment:
can you negate and return and reduce the nesting block?

Reply:
Done.